src/bottom_topography/spline_cache_2D.jl

```diff
@@ -0,0 +1,360 @@
+  IP::IP_type
+
+  BilinearBSpline(x, y, z, h, Q, IP) = new{typeof(x), typeof(y), typeof(z),
+  typeof(h), typeof(Q), typeof(IP)}(x, y, z, h, Q, IP)
+end
+
+# Fill structure
+function bilinear_b_spline(x::Vector, y::Vector, z::Matrix; smoothing_factor = 0.0)
+
+  x, y, z = sort_data(x,y,z)
+
+  # Consider spline smoothing if required
+  if smoothing_factor > 0.0
+    z = calc_tps(smoothing_factor, x, y, z)
+  end
+
+  n = length(x)
+  m = length(y)
+  h = x[2] - x[1]
+  
+  P = vcat(reshape(z', (m*n,1)))
+  IP = [-1 1;
+         1 0];
+
+  Q = reshape(P, (n, m)) 
+
+  BilinearBSpline(x, y, z, h, Q, IP)  
+end
+
+# Read from file
+function bilinear_b_spline(path::String; smoothing_factor = 0.0)
+  file = open(path)
+  lines = readlines(file)
+  close(file)
+
+  n = parse(Int64, lines[2])
+  m = parse(Int64, lines[4])
+  
+  x     = [parse(Float64, val) for val in lines[6      :5+n    ]]
+  y     = [parse(Float64, val) for val in lines[(7+n)  :(6+n+m)]]
+  z_tmp = [parse(Float64, val) for val in lines[(8+n+m):end    ]]
+
+  z = transpose(reshape(z_tmp, (n, m)))
+
+  bilinear_b_spline(x, y, Matrix(z); smoothing_factor = smoothing_factor)
+end
+
+######################################
+### Bicubic B Spline interpolation ###
+######################################
+
+# Bicubic B Spline structure
+# The entries of z are set to be the x values horizontally
+# and y values vertically
+mutable struct BicubicBSpline{x_type, y_type, z_type, h_type, Q_type, IP_type}
+
+  x::x_type
+  y::y_type
+  z::z_type
+  h::h_type
+  Q::Q_type
+  IP::IP_type
+
+  BicubicBSpline(x, y, z, h, Q, IP) = new{typeof(x), typeof(y), typeof(z), typeof(h), 
+  typeof(Q), typeof(IP)}(x, y, z, h, Q, IP)
+end
+
+# Fill structure
+function bicubic_b_spline(x::Vector, y::Vector, z::Matrix; boundary = "free", smoothing_factor = 0.0)
+
+  x, y, z = sort_data(x,y,z)
+
+  # Consider spline smoothing if required
+  if smoothing_factor > 0.0
+    z = calc_tps(smoothing_factor, x, y, z)
+  end
+
+  n = length(x)
+  m = length(y)
+  h = x[2] - x[1]
+  boundary_elmts = 4 + 2*m + 2*n
+  inner_elmts = m*n
+  P = vcat(reshape(z', (inner_elmts,1)), zeros(boundary_elmts))
+  
+  IP = [-1  3 -3 1;
+         3 -6  3 0;
+        -3  0  3 0;
+         1  4  1 0]
+
+  # Fill Phi for the different boundary conditions
+
+  ########################
+  ## Free end condition ##
+  ########################
+  if boundary == "free"
+    Phi = spzeros((m+2)*(n+2), (m+2)*(n+2))
+
+    # Fill inner point matrix
+    idx = 0
+    for i in 1:inner_elmts
+      Phi[i, idx           + 1] =  1
+      Phi[i, idx           + 2] =  4
+      Phi[i, idx           + 3] =  1
+      Phi[i, idx +   (n+2) + 1] =  4
+      Phi[i, idx +   (n+2) + 2] = 16
+      Phi[i, idx +   (n+2) + 3] =  4
+      Phi[i, idx + 2*(n+2) + 1] =  1
+      Phi[i, idx + 2*(n+2) + 2] =  4
+      Phi[i, idx + 2*(n+2) + 3] =  1
+
+      if (i % n) == 0
+        idx += 3
+      else
+        idx += 1
+      end
+    end
+
+    # left edge (v-)
```

Comment:
It would be good to put the ansatz you are using somewhere so it is more clear to a user what the v± and u± mean.

Reply:
Has been documented